apis/traffic-access/v1alpha4/traffic-access.md

```diff
@@ -0,0 +1,352 @@
+# Traffic Access Control
+
+**API Group:** access.smi-spec.io
+
+**API Version:** v1alpha4
+
+**Compatible with:** specs.smi-spec.io/v1alpha4
+
+## Specification
+
+This set of resources allows users to define access control policy for their
+applications. It is the authorization side of the picture. Authentication should
+already be handled by the underlying implementation and surfaced through a subject.
+
+Access control in this specification is additive, all traffic is denied by default.
+See [tradeoffs](#tradeoffs) for a longer discussion about why.
+
+### IdentityBinding
+
+An `IdentityBinding` declares the set of identities belonging to a particular workload
+for the purposes of policy. At present, Kubernetes does not natively provide any
+sort of identity resource outside of `ServiceAccount`. As such, many mesh
+implementations have turned to alternative identification schemes for more control
+over traffic routign and policy (e.g. SPIFFE, pod selectors, etc). Unfortunately,
+these arbitrary identity mechanisms are rarely stored in a machine-accessible
+manner. This is the role of the `IdentityBinding` resource.
+
+```yaml
+apiVersion: access.smi-spec.io/v1alpha4
+kind: IdentityBinding
+metadata:
+ name: service-a
+ namespace: default
+spec:
+ schemes:
+   podLabelSelectors:
```

Comment:
Given that pods must run under a service account, I don’t think we should allow pod selectors here, the reason for this is that we would bind an SA to a pod that runs under a different one and that conflicts with Kubernetes. Another issue with this proposal is that users could bind the same pod to multiple SAs, yet again this goes against  Kubernetes workload identity. I propose we limit the `IdentityBinding` scope to binding an SA to some external identity, we should also make the service account field mandatory.

Reply:
Excellent point! Part of the impetus for this new resource is to allow for interop with Istio [AuthorizationPolicies](https://istio.io/latest/docs/reference/config/security/authorization-policy/#AuthorizationPolicy) that only accept a label selector to limit the scope of the access policy. I do agree with you however  that declaring a service account and a selector inherently contradict. I think I would propose the opposite of your suggestion: make service account and podlabelselector mutually exclusive, enabling cluster operators to create a sort of weak binding to arbitrary workloads based on labels, (exclusive) OR they can use k8s baked in strong identity with service accounts and all of their guarantees to bind to external identities. What do you think?

Reply:
Is there a need for `spiffeIdentities` to be AND-able with either `serviceAccount` or `podLabelSelectors`? Consul supports both SPIFFE identifiers and service accounts, but I think specifying either alone would be sufficient and if that is the that case maybe only a single `scheme` should be supported?
/cc @nicholasjackson

Reply:
I'm a bit wary of relying solely on service accounts for SPIFFE IDs since according to the [spec](https://spiffe.io/docs/latest/spiffe-about/spiffe-concepts/#spiffe-id), they can be within any trust domain. A user running SPIRE with a custom trust domain (or federated trust domains) would need more explicit control over the IDs. I think the fact that the schemes are OR'd (i.e a matching workload can fit scheme A OR B) allows controllers (like Consul) to limit which schemes they support

Reply:
>  I think I would propose the opposite of your suggestion: make service account and podlabelselector mutually exclusive, enabling cluster operators to create a sort of weak binding to arbitrary workloads based on labels, (exclusive) OR they can use k8s baked in strong identity with service accounts and all of their guarantees to bind to external identities. What do you think?

Yes I think making service account and podlabelselector mutually exclusive would be best. We can't enforce this in the SDK because controller-gen doesn't support OR’d fields, but this validation could be implemented in an admission controller.